mne/preprocessing/ica.py

```diff
@@ -660,29 +660,14 @@ def _compute_pre_whitener(self, data):
             # Scale (z-score) the data by channel type
             info = self.info
             pre_whitener = np.empty([len(data), 1])
-            for ch_type in _DATA_CH_TYPES_SPLIT + ('eog', "ref_meg"):
-                if _contains_ch_type(info, ch_type):
-                    if ch_type == 'seeg':
-                        this_picks = pick_types(info, meg=False, seeg=True)
-                    elif ch_type == 'dbs':
-                        this_picks = pick_types(info, meg=False, dbs=True)
-                    elif ch_type == 'ecog':
-                        this_picks = pick_types(info, meg=False, ecog=True)
-                    elif ch_type == 'eeg':
-                        this_picks = pick_types(info, meg=False, eeg=True)
-                    elif ch_type in ('mag', 'grad'):
-                        this_picks = pick_types(info, meg=ch_type)
-                    elif ch_type == 'eog':
-                        this_picks = pick_types(info, meg=False, eog=True)
-                    elif ch_type in ('hbo', 'hbr'):
-                        this_picks = pick_types(info, meg=False, fnirs=ch_type)
-                    elif ch_type == 'ref_meg':
-                        this_picks = pick_types(info, meg=False, ref_meg=True)
-                    else:
-                        raise RuntimeError('Should not be reached.'
-                                           'Unsupported channel {}'
-                                           .format(ch_type))
-                    pre_whitener[this_picks] = np.std(data[this_picks])
+            for _, picks_ in _picks_by_type(info, ref_meg=False, exclude=[]):
+                pre_whitener[picks_] = np.std(data[picks_])
+            if _contains_ch_type(info, "ref_meg"):
+                picks_ = pick_types(info, ref_meg=True, exclude=[])
+                pre_whitener[picks_] = np.std(data[picks_])
+            if _contains_ch_type(info, "eog"):
+                picks_ = pick_types(info, eog=True, exclude=[])
+                pre_whitener[picks_] = np.std(data[picks_])
```

Comment:
why do you need to special case this? This suggests there is something inconsistent somewhere.

Reply:
First of all, they were already both special cases before. 
Second, the problem is, as I started pointing out in a comment above, in `_picks_by_type` which picks only data channels (this is not a bug, this is what this function is intended for).

Example code:
```
import mne

ch_names = [f'EEG{i+1}' for i in range(4)] + \
           ['EOG1'] + ['GRAD1', 'GRAD2', 'MAG1'] + ['REF_MEG']
ch_types = ['eeg'] * 4 + ['eog'] + ['grad'] * 2 + ['mag'] + ['ref_meg']
info = mne.create_info(ch_names, ch_types=ch_types, sfreq=1000)
```

Missing the EOG channel:
```
_picks_by_type(info)
Out: [('mag', array([7])), ('grad', array([5, 6])), ('eeg', array([0, 1, 2, 3]))]
```

Adding the MEG REF as a MAG: **Bug?**
```
_picks_by_type(info, ref_meg=True)
Out: [('mag', array([7, 8])), ('grad', array([5, 6])), ('eeg', array([0, 1, 2, 3]))]
```

Reply:
>> This suggests there is something inconsistent somewhere.

> Second, the problem is, as I started pointing out in a comment above, in _picks_by_type which picks only data channels (this is not a bug, this is what this function is intended for).

IIRC the whitening function does not usually operate on EOG or ref_mag, but for ICA we want to include them. So indeed I agree what you have here is one option.

However, I'm not sure `compute_whitener` can *only* operate on these. Can you pass all `picks` to it as `compute_whitener(..., picks=all_data_plus_eog_and_ref_mag_picks)` and have it do the right thing? It might work...

Reply:
I don't follow. The changes here I limited to the case where the noise covariance matrix is not provided, and thus the case where `compute_whitener` is not called.
Note that I didn't test the case where the noise covariance matrix is provided. By the code in `compute_whitener`, I guess it will fail if you include bad channels as picks are handled with: `_picks_to_idx(info, picks, with_ref_meg=False)`.

Reply:
> The changes here I limited to the case where the noise covariance matrix is not provided

Oh right, never mind then!

Reply:
@larsoner In the end.. following your comment.. I did add and fixed as well the part with the covariance matrix provided.